src/lib.rs

```diff
@@ -0,0 +1,150 @@
+//! A storage device representation to help filesystem development.
+//! Designed to be used in filesystem drivers.
+
+#![no_std]
+
+/// Block device representation.
+pub mod block;
+
+pub use block::*;
+
+/// Represent a storage device error.
+#[derive(Debug)]
+pub enum StorageDeviceError {
+    /// Read error
+    ReadError,
+
+    /// Write error
+    WriteError,
+
+    /// Unknown error
+    Unknown,
+}
+
+/// Represent a storage device result.
+pub type StorageDeviceResult<T> = core::result::Result<T, StorageDeviceError>;
+
+/// Represent a device managing storage.
+// we don't need is_empty, this would be stupid.
+#[allow(clippy::len_without_is_empty)]
+pub trait StorageDevice {
+    /// Read the data at the given ``offset`` in the storage device into a given buffer.
+    fn read(&self, offset: u64, buf: &mut [u8]) -> StorageDeviceResult<()>;
+
+    /// Write the data from the given buffer at the given ``offset`` in the storage device.
+    fn write(&self, offset: u64, buf: &[u8]) -> StorageDeviceResult<()>;
+
+    /// Return the total size of the storage device.
+    fn len(&self) -> StorageDeviceResult<u64>;
+}
+
+impl From<BlockError> for StorageDeviceError {
+    fn from(error: BlockError) -> Self {
+        match error {
+            BlockError::ReadError => StorageDeviceError::ReadError,
+            BlockError::WriteError => StorageDeviceError::WriteError,
+            BlockError::Unknown => StorageDeviceError::Unknown,
+        }
+    }
+}
+
+/// Implementation of storage device for block device.
+pub struct StorageBlockDevice<B: BlockDevice> {
+    /// The inner block device.
+    block_device: B,
+}
+
+impl<B: BlockDevice> StorageBlockDevice<B> {
+    /// Create a new storage block device.
+    pub fn new(block_device: B) -> Self {
+        StorageBlockDevice { block_device }
+    }
+}
+
+impl<B: BlockDevice> StorageDevice for StorageBlockDevice<B> {
+    fn read(&self, offset: u64, buf: &mut [u8]) -> StorageDeviceResult<()> {
+        let mut read_size = 0u64;
+        let mut blocks = [Block::new()];
```

Comment:
Reading 1 block at a time from disk has terrible performance. You should allocate the blocks, perform **one** read of all the concerned blocks from the wrapped BlockDevice, and copy the interesting part into the returned slice.